File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D18.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Grid
import org.gristle.pdxpuzzles.utilities.objects.toGrid

class Y15D18(input: String) : Day {

    private val lights = input.toGrid { it == '#' }

    private fun Grid<Boolean>.iterate(cornersStuck: Boolean = false): Grid<Boolean> {
        val corners = if (cornersStuck) {
            listOf(0, width - 1, size - width, lastIndex)
        } else {
            emptyList()
        }

        return Grid(height, width) { i ->
            if (i in corners) {
                true
            } else {
                val neighborsOn = getNeighbors(i, true).count { it }
                neighborsOn == 3 || (neighborsOn == 2 && this[i])
            }
        }
    }

    private fun solve(cornersStuck: Boolean) = generateSequence(lights) { it.iterate(cornersStuck) }
        .take(101)
        .last()
        .count { it }

    override fun part1() = solve(false)

    override fun part2() = solve(true)
}

fun main() = Day.runDay(Y15D18::class)

//    Class creation: 24ms
//    Part 1: 1061 (338ms)
//    Part 2: 1006 (218ms)
//    Total time: 581ms

File: rust/advent/src/bin/2015/y15d18.rs
use std::iter::successors;

use advent::utilities::get_input::get_input;
use utilities::structs::{grid::Grid2, stopwatch::{ReportDuration, Stopwatch}};

type Input = Grid2<bool>;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 18).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    Grid2::new2d_map_str(input, |c| c == '#').unwrap()
}

fn iterate(lights: &Input, corners_stuck: bool) -> Input {
    let corners: Option<[usize; 4]> = if corners_stuck {
        Some([0, lights.width() - 1, lights.len() - lights.width(), lights.len() - 1])
    } else {
        None
    };

    Grid2::new2d_with_fn(lights.width(), lights.height(), |i| {
        if let Some(corners) = corners && corners.contains(&i) {
            true
        } else {
            let neighbors_on = lights
                .adjacent(i, true)
                .unwrap()
                .filter(|adj| *adj.value)
                .count();
            neighbors_on == 3 || (neighbors_on == 2 && lights[i])
        }
    })
}

fn solve(lights: &Input, corners_stuck: bool) -> Output {
    successors(Some(lights.clone()), |it| Some(iterate(it, corners_stuck)))
        .take(101)
        .last()
        .unwrap()
        .iter()
        .filter(|&&it| it)
        .count()
}

fn part1(lights: &Input) -> Output {
    solve(lights, false)
}

fn part2(lights: &Input) -> Output {
    solve(lights, true)
}

#[test]
fn default() {
    let input = get_input(15, 18).unwrap();
    let input = parse_input(&input);
    assert_eq!(1061, part1(&input));
    assert_eq!(1006, part2(&input));
}

// Input parsed (72μs)
// 1. 1061 (108ms)
// 2. 1006 (106ms)
// Total: 214ms
